src/lib/operators/insert.cpp

```diff
@@ -8,94 +8,80 @@
 #include "resolve_type.hpp"
 #include "storage/base_encoded_segment.hpp"
+#include "storage/segment_iterate.hpp"
 #include "storage/storage_manager.hpp"
 #include "storage/value_segment.hpp"
 #include "type_cast.hpp"
 #include "utils/assert.hpp"
 
-namespace opossum {
+namespace {
 
-// We need these classes to perform the dynamic cast into a templated ValueSegment
-class AbstractTypedSegmentProcessor : public Noncopyable {
- public:
-  AbstractTypedSegmentProcessor() = default;
-  AbstractTypedSegmentProcessor(const AbstractTypedSegmentProcessor&) = delete;
-  AbstractTypedSegmentProcessor& operator=(const AbstractTypedSegmentProcessor&) = delete;
-  AbstractTypedSegmentProcessor(AbstractTypedSegmentProcessor&&) = default;
-  AbstractTypedSegmentProcessor& operator=(AbstractTypedSegmentProcessor&&) = default;
-  virtual ~AbstractTypedSegmentProcessor() = default;
-  virtual void resize_vector(std::shared_ptr<BaseSegment> segment, size_t new_size) = 0;
-  virtual void copy_data(std::shared_ptr<const BaseSegment> source, ChunkOffset source_start_index,
-                         std::shared_ptr<BaseSegment> target, ChunkOffset target_start_index, ChunkOffset length) = 0;
-};
+using namespace opossum;  // NOLINT
 
 template <typename T>
-class TypedSegmentProcessor : public AbstractTypedSegmentProcessor {
- public:
-  void resize_vector(std::shared_ptr<BaseSegment> segment, size_t new_size) override {
-    auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
-    DebugAssert(value_segment, "Cannot insert into non-ValueColumns");
-    auto& values = value_segment->values();
+void resize_value_segment(std::shared_ptr<BaseSegment> segment, size_t new_size) {
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
+  Assert(value_segment, "Cannot insert into non-ValueColumns");
 
-    values.resize(new_size);
+  value_segment->values().resize(new_size);
 
-    if (value_segment->is_nullable()) {
-      value_segment->null_values().resize(new_size);
-    }
+  if (value_segment->is_nullable()) {
+    value_segment->null_values().resize(new_size);
   }
+}
 
-  // this copies
-  void copy_data(std::shared_ptr<const BaseSegment> source, ChunkOffset source_start_index,
-                 std::shared_ptr<BaseSegment> target, ChunkOffset target_start_index, ChunkOffset length) override {
-    auto casted_target = std::dynamic_pointer_cast<ValueSegment<T>>(target);
-    DebugAssert(casted_target, "Cannot insert into non-ValueColumns");
-    auto& values = casted_target->values();
-
-    auto target_is_nullable = casted_target->is_nullable();
-
-    if (auto casted_source = std::dynamic_pointer_cast<const ValueSegment<T>>(source)) {
-      std::copy_n(casted_source->values().begin() + source_start_index, length, values.begin() + target_start_index);
+template <typename T>
+void copy(const std::shared_ptr<const BaseSegment>& source_base_segment, ChunkOffset source_begin_offset,
```

Comment:
copy what? I agree that data doesn't mean too much either. copy attribute? That's what the relational algebra would call it.

Reply:
hard one to name right, it is very similar to, e.g., `std::copy`/`memcpy` - just that it operates on segments. `copy_value_range`?

Reply:
I like it